In [1]:
import pandas as pd
import numpy as np

df = pd.read_excel('Base Etudiant.xls')
df = df.drop_duplicates()
df.head(20)

id_et         date_nais_et lieu_nais_et             date_bac  \
0    ID1  1996-07-10 00:00:00        Tunis                  NaN   
1    ID2  2002-12-19 00:00:00        TUNIS  2021-06-01 00:00:00   
2    ID3  2003-08-09 00:00:00        TUNIS  2024-06-01 00:00:00   
3    ID4  1999-09-15 00:00:00        TUNIS                  NaN   
4    ID5  1999-10-17 00:00:00      Bizerte  2020-06-01 00:00:00   
5    ID6  2004-07-24 00:00:00        tunis  2024-06-01 00:00:00   
6    ID7  2004-03-24 00:00:00        Tunis  2024-06-01 00:00:00   
7    ID8  2002-09-09 00:00:00        Gafsa  2022-06-01 00:00:00   
8    ID9  2003-06-12 00:00:00   Benguerden  2022-06-01 00:00:00   
9   ID10  2004-11-06 00:00:00         SFAX  2023-06-01 00:00:00   
10  ID11  1999-08-12 00:00:00        TUNIS                  NaN   
11  ID12  2000-09-13 00:00:00        TUNIS                  NaN   
12  ID13  2004-01-17 00:00:00       Menzah  2022-06-01 00:00:00   
13  ID14  2001-11-06 00:00:00         Sfax  2020-08-09 00:00:00   
14  ID15  2001-10-01 00:00:00        tunis  2020-06-01 00:00:00   
15  ID16  2001-12-06 00:00:00     MEDENINE                  NaN   
16  ID17  1999-05-16 00:00:00      KELIBIA                  NaN   
17  ID18  2003-05-13 00:00:00        Gabès  2022-06-01 00:00:00   
18  ID19  2005-09-22 00:00:00        Tunis  2024-06-01 00:00:00   
19  ID20  2004-09-02 00:00:00         Sfax  2023-06-01 00:00:00   

                           etab_bac  niveau_diplome_sup_et  \
0                               NaN                    NaN   
1         Lycée Secondaire Ezzahra                     3.0   
2   Lycee Secondaire ASDRUBAL BARDO                    0.0   
3            Lycée les pères blancs                    NaN   
4               Lycée Bhira Bizerte                    0.0   
5                        el ghazela                    0.0   
6      Lycée El Ahdeth El Mourouj 6                    0.0   
7                            Nabeul                    0.0   
8                         bengarden                    0.0   
9                 Lycée Borj Cedria                    0.0   
10     lycée pilote bourguiba Tunis                    NaN   
11                              NaN                    NaN   
12                        Carthage                     0.0   
13        Lycée 14 janvier - Sousse                    0.0   
14                            tunis                    0.0   
15        Lycée technique medenine                     NaN   
16       Lyceé Abdelaziz el khouja                     NaN   
17                            gabes                    0.0   
18                 Lycée Pere Blanc                    0.0   
19  Lycéc Mohamed BOUDHINA Hammamet                    1.0   

                                lib_specialite_esp_et   date_entree_esp_et  \
0         Master Professionnel en Business Analytics                   NaN   
1           Master Professionnel en Marketing Digital  2024-09-16 00:00:00   
2                      Licence en Sciences de Gestion  2024-08-30 00:00:00   
3   Master Professionnel en Management Digital et ...                  NaN   
4                      Licence en Sciences de Gestion  2023-09-18 00:00:00   
5                      Licence en Sciences de Gestion  2024-09-05 00:00:00   
6                          Licence Business Computing  2024-08-22 00:00:00   
7   Licence Business Computing- Parcours  Business...                  NaN   
8   Licence Business Computing- Parcours  Business...                  NaN   
9                         Licence Business Computing   2024-09-03 00:00:00   
10          Master Professionnel  Business Analytics                   NaN   
11  Master Professionnel en Management  Digital et...                  NaN   
12                         Licence Business Computing  2023-09-01 00:00:00   
13                         Licence Business Computing  2023-08-25 00:00:00   
14  Licence en Sciences de Gestion -Parcours Compt...                  NaN   
15                         Licence Business 

In [2]:
df.shape

(2621, 20)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2621 entries, 0 to 2620
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id_et                  2621 non-null   object 
 1   date_nais_et           2621 non-null   object 
 2   lieu_nais_et           2619 non-null   object 
 3   date_bac               2259 non-null   object 
 4   etab_bac               2457 non-null   object 
 5   niveau_diplome_sup_et  2259 non-null   float64
 6   lib_specialite_esp_et  2621 non-null   object 
 7   date_entree_esp_et     1388 non-null   object 
 8   lib_nature_bac         2563 non-null   object 
 9   classe_courante_et     2621 non-null   object 
 10  classe_sem1            2369 non-null   object 
 11  niveau_courant_et      2621 non-null   int64  
 12  sexe                   2621 non-null   object 
 13  nationalite            2619 non-null   object 
 14  niveau_acces           2621 non-null   int64  
 15  clas

# Update column names for clarity

In [4]:
df.columns

Index(['id_et', 'date_nais_et', 'lieu_nais_et', 'date_bac', 'etab_bac',
       'niveau_diplome_sup_et', 'lib_specialite_esp_et', 'date_entree_esp_et',
       'lib_nature_bac', 'classe_courante_et', 'classe_sem1',
       'niveau_courant_et', 'sexe', 'nationalite', 'niveau_acces',
       'classe_prec_et', 'resultat_annee_prec', 'moy_bac', 'score_final',
       'gouvernorat'],
      dtype='object')

In [5]:
column_mapping = {
    'id_et': 'ID',
    'date_nais_et': 'Birth date',
    'lieu_nais_et': 'Birth place',
    'date_bac': 'Highschool graduation date',
    'etab_bac': 'Highschool',
    'niveau_diplome_sup_et': 'Studies level',
    'lib_specialite_esp_et': 'Degree',
    'date_entree_esp_et': 'ESB entry date',
    'lib_nature_bac': 'Bac type',
    'classe_courante_et': 'Current class',
    'classe_sem1': 'Semester 1 class',
    'niveau_courant_et': 'Degree level',
    'sexe': 'Gender',
    'nationalite': 'Nationality',
    'niveau_acces': 'Class level',
    'classe_prec_et': 'Previous class',
    'resultat_annee_prec': 'Previous year result',
    'moy_bac': 'Highschool graduation grade',
    'score_final': 'Final score',
    'gouvernorat': 'Region'
}

df = df.rename(columns=column_mapping)
df.head(1)

ID           Birth date Birth place Highschool graduation date Highschool  \
0  ID1  1996-07-10 00:00:00       Tunis                        NaN        NaN   

   Studies level                                       Degree ESB entry date  \
0            NaN  Master Professionnel en Business Analytics             NaN   

  Bac type Current class Semester 1 class  Degree level Gender Nationality  \
0      NaN        2-BA-2           2-BA-2             2      M  TUNISIENNE   

   Class level Previous class Previous year result  \
0            5         1-BA-2               Admis    

   Highschool graduation grade  Final score Region  
0                          NaN         -1.0    NaN

# String values consistency

In [6]:
df['Studies level'].value_counts()

Studies level
0.0    1684
3.0     492
1.0      81
2.0       2
Name: count, dtype: int64

In [7]:
df['Previous year result'].value_counts()

Previous year result
ADMIS    Mention Passable         355
ADMISE   Mention Passable         312
ADMISE   Mention Assez Bien       247
ADMIS    Mention Assez Bien       180
Admis                             106
ADMISE   Mention Bien              98
REDOUBLE                           91
Admise                             54
ADMIS    Mention Bien              35
ADMIS  Avec 10 ECTS non acquis     29
ADMISE Avec 10 ECTS non acquis     20
ADMIS  Avec 6 ECTS non acquis      20
ADMIS  Avec 12 ECTS non acquis     13
ADMIS                              13
ADMIS  Avec 15 ECTS non acquis     13
ADMISE   Mention Très Bien         11
ADMIS  Avec 4 ECTS non acquis       9
ADMISE Avec 15 ECTS non acquis      9
EXCLU                               9
ADMIS  Avec 5 ECTS non acquis       8
ADMIS  Avec 8 ECTS non acquis       8
ADMISE Avec 12 ECTS non acquis      7
ADMISE Avec 6 ECTS non acquis       7
ADMIS  Avec 14 ECTS non acquis      6
ADMISE                              5
ADMISE Avec 14 ECTS non acqui

In [8]:
# Apply string standardization to the columns to group similar values together

#birth place column:
df['Birth place'] = df['Birth place'].str.strip().str.lower().str.title()

#previous year result column
df['Previous year result'] = df['Previous year result'].str.strip().str.lower().str.title()

# Apply consistenct map across values for previous year result values. Using str.contains to look for matching values
mappingRes = {
    'Passable': 'Admis(e) - Passable',
    'Assez Bien': 'Admis(e) - Assez Bien',
    'Bien': 'Admis(e) - Bien',
    'Très Bien': 'Admis(e) - Très Bien',
    'Ects': 'Admis(e) - Missing credits',
    'Ex': 'Exclu(e)',
    'Redouble': 'Redouble'
}

for old, replace in mappingRes.items():
    df.loc[df['Previous year result'].str.contains(old, na=False), 'Previous year result'] = replace

df['Previous year result'] = df['Previous year result'].replace(to_replace=['Admis', 'Admise'], value='Admis(e)'
)

In [9]:

# ensure consistency across values for degree values

df['Degree'] = df['Degree'].str.strip().str.lower().str.title()

mappingRes = {
    'Computing': 'LBC',
    'Gestion': 'LSG',
    'Analytics': 'BA',
    'Management': 'MDSI',
    'Marketing': 'MKD',
    'Finance': 'FIND',
    'Audit': 'CCA',
    'Décision': 'LMAD',
    'Actuarielle': 'GAMMA',
    
}

for old, replace in mappingRes.items():
    df.loc[df['Degree'].str.contains(old, na=False), 'Degree'] = replace



df['Degree'].value_counts()



Degree
LBC     1142
LSG      653
BA       397
MDSI     249
MKD       89
FIND      54
CCA       29
LMAD       8
Name: count, dtype: int64

In [10]:


# ensure consistency across values for bac type values

df['Bac type'] = df['Bac type'].str.strip().str.lower().str.title()

mappingRes = {
    'Eco': 'Bac Economie',
    'Sci': 'Bac Sciences',
    'Math': 'Bac Math',
    'Téch|Tech': 'Bac Technique',
    'Lettre': 'Bac Lettres',
    'ra': 'Bac Français'
    
}

for old, replace in mappingRes.items():
    df.loc[df['Bac type'].str.contains(old, na=False), 'Bac type'] = replace



df['Bac type'].value_counts()




Bac type
Bac Economie        1521
Bac Sciences         511
Bac Math             260
Bac Technique        142
Bac Informatique     119
Autre Bac              5
Bac Français           4
Bac Lettres            1
Name: count, dtype: int64

In [11]:
df['Gender'].unique()

array(['M', 'F', 'H'], dtype=object)

In [12]:
# Replace 'H' with 'M' in Gender column
df['Gender'] = df['Gender'].replace('H', 'M')

In [13]:
# Convert date columns to datetime format, only keeep the year, then convert back to integer; end result Example: 2024 instead of 2024-12-12 00:00:00:00
df['Birth date'] = pd.to_datetime(df['Birth date'], errors='coerce').dt.year.astype('Int64')
df['Highschool graduation date'] = pd.to_datetime(df['Highschool graduation date'], errors='coerce').dt.year.astype('Int64')
df['ESB entry date'] = pd.to_datetime(df['ESB entry date'], errors='coerce').dt.year.astype('Int64')

In [14]:
# Apply string standardization to the Highschool column
df['Highschool'] = df['Highschool'].str.strip().str.lower().str.title()

# Replace specific substrings
df['Highschool'] = df['Highschool'].str.replace(' Secondaire', '', regex=False)
df['Highschool'] = df['Highschool'].str.replace('Lycée', 'Lycee', regex=False)

# Handle invalid or missing values
df['Highschool'] = df['Highschool'].apply(lambda x: x if isinstance(x, str) else np.nan)

df['Highschool'] = df['Highschool'].apply(
    lambda x: 'Lycee ' + x if isinstance(x, str) and 'Lycee' not in x else x
)



In [15]:

# Display the unique values to verify the cleaning
df['Highschool'].value_counts()

Highschool
Lycee Tunis                    235
Lycee Ariana                   158
Lycee Bizerte                   53
Lycee Nabeul                    39
Lycee Menzah 6                  34
                              ... 
Lycee Genève                     1
Lycee Le Sers El Kef             1
Lycee Hamouda Becha Manouba      1
Lycee Menzel Tmime               1
Lycee Hammam Ghzez               1
Name: count, Length: 1082, dtype: int64

In [16]:
df

ID  Birth date Birth place  Highschool graduation date  \
0        ID1        1996       Tunis                        <NA>   
1        ID2        2002       Tunis                        2021   
2        ID3        2003       Tunis                        2024   
3        ID4        1999       Tunis                        <NA>   
4        ID5        1999     Bizerte                        2020   
...      ...         ...         ...                         ...   
2616  ID2617        2000       Tunis                        2022   
2617  ID2618        2005      Le Kef                        2023   
2618  ID2619        2004    La Marsa                        2023   
2619  ID2620        2004       Tunis                        2022   
2620  ID2621        2003     Bizerte                        2023   

                  Highschool  Studies level Degree  ESB entry date  \
0                        NaN            NaN     BA            <NA>   
1              Lycee Ezzahra            3.0    MKD            2024   
2       Lycee Asdrubal Bardo            0.0    LSG            2024   
3     Lycee Les Pères Blancs            NaN   MDSI            <NA>   
4        Lycee Bhira Bizerte            0.0    LSG            2023   
...                      ...            ...    ...             ...   
2616           Lycee Tunis 2            0.0    LBC            <NA>   
2617               Lycee Kef            0.0    LBC            <NA>   
2618          Lycee L'Aouina            0.0    LSG            2023   
2619          Lycee La Marsa            0.0    LSG            <NA>   
2620           Lycee Bizerte            0.0    LBC            <NA>   

              Bac type Current class Semester 1 class  Degree level Gender  \
0                  NaN        2-BA-2           2-BA-2             2      M   
1     Bac Informatique       1-MKD-1              ---             1      M   
2         Bac Economie       1-LSG-1          1-LSG-1             1      M   
3         Bac Economie      1-MDSI-1         1-MDSI-1             1      F   
4         Bac Economie       2-LSG-3          2-LSG-3             2      M   
...                ...           ...              ...           ...    ...   
2616      Bac Economie    3LBC-BIS-5       3LBC-BIS-5             3      M   
2617      Bac Sciences       2-LBC-6          2-LBC-6             2      M   
2618      Bac Economie       2-LSG-3          2-LSG-3             2      M   
2619      Bac Sciences       3-LSC-2          3-LSC-2             3      F   
2620      Bac Economie       2-LBC-8          2-LBC-8             2      F   

     Nationality  Class level Previous class        Previous year result  \
0     TUNISIENNE            5         1-BA-2                    Admis(e)   
1     TUNISIENNE            4            NaN                         NaN   
2     TUNISIENNE            1            NaN                         NaN   
3     TUNISIENNE            4        3-LSG-4             Admis(e) - Bien   
4     TUNISIENNE            2        1-LSG-8         Admis(e) - Passable   
...          ...          ...            ...                         ...   
2616  TUNISIENNE            3     2-LBC-BIS5         Admis(e) - Passable   
2617  TUNISIENNE            2        1-LBC-9             Admis(e) - Bien   
2618  TUNISIENNE            2        1-LSG-3  Admis(e) - Missing credits   
2619  TUNISIENNE            3        2-LSC-2             Admis(e) - Bien   
2620  TUNISIENNE            2       1-LBC-11         Admis(e) - Passable   

      Highschool graduation grade  Final score     Region  
0                             NaN         -1.0        NaN  
1                           12.91         -1.0  Ben Arous  
2                            9.81         -1.0     Ariana  
3                           12.00         -1.0        NaN  
4                           10.48         -1.0    Bizerte  
...                           ...          ...        ...  
2616                        11.12         -1.0     Ariana  
2617                  

In [17]:
# Fields in columns 'Previous class', 'Semester 1 class', 'Previous year result' have '---' values. Replace them with NaN
df[['Previous class', 'Semester 1 class', 'Previous year result']] = df[['Previous class', 'Semester 1 class', 'Previous year result']].replace('---', np.nan)

In [18]:
df['Region'].nunique()

24

# Filling null values

In [19]:
#Check if there are rows where entry date can't be determined based on highschool graduation
df[(df['ESB entry date'].isna()) & (df['Highschool graduation date'].isna())]

ID  Birth date Birth place  Highschool graduation date  \
0        ID1        1996       Tunis                        <NA>   
3        ID4        1999       Tunis                        <NA>   
10      ID11        1999       Tunis                        <NA>   
11      ID12        2000       Tunis                        <NA>   
15      ID16        2001    Medenine                        <NA>   
...      ...         ...         ...                         ...   
2603  ID2604        2002       Tunis                        <NA>   
2610  ID2611        1999        Sfax                        <NA>   
2611  ID2612        2002    Menzah 6                        <NA>   
2613  ID2614        2000    Menzah 6                        <NA>   
2615  ID2616        2001       Tunis                        <NA>   

                        Highschool  Studies level Degree  ESB entry date  \
0                              NaN            NaN     BA            <NA>   
3           Lycee Les Pères Blancs            NaN   MDSI            <NA>   
10    Lycee Pilote Bourguiba Tunis            NaN     BA            <NA>   
11                             NaN            NaN   MDSI            <NA>   
15        Lycee Technique Medenine            NaN    LBC            <NA>   
...                            ...            ...    ...             ...   
2603                Lycee Menzah 6            NaN    LSG            <NA>   
2610           Lycee Habib Thameur            NaN   MDSI            <NA>   
2611                           NaN            NaN     BA            <NA>   
2613                           NaN            NaN     BA            <NA>   
2615                           NaN            NaN     BA            <NA>   

           Bac type Current class Semester 1 class  Degree level Gender  \
0               NaN        2-BA-2           2-BA-2             2      M   
3      Bac Economie      1-MDSI-1         1-MDSI-1             1      F   
10     Bac Sciences        1-BA-8              NaN             1      M   
11         Bac Math      2-MDSI-4         2-MDSI-4             2      F   
15    Bac Technique       2-LBC-1          2-LBC-1             2      F   
...             ...           ...              ...           ...    ...   
2603   Bac Economie       3-LSG-5          3-LSG-5             3      M   
2610   Bac Sciences      2-MDSI-6              NaN             2      F   
2611   Bac Economie        1-BA-8              NaN             1      F   
2613   Bac Economie        2-BA-4           2-BA-4             2      M   
2615   Bac Economie        2-BA-5           2-BA-5             2      F   

     Nationality  Class level Previous class        Previous year result  \
0     TUNISIENNE            5         1-BA-2                    Admis(e)   
3     TUNISIENNE            4        3-LSG-4             Admis(e) - Bien   
10    TUNISIENNE            4     1DSI-AS2-1             Admis(e) - Bien   
11    TUNISIENNE            5       1-MDSI-4         Admis(e) - Passable   
15    TUNISIENNE            2     2-LBC-BIS4                    Exclu(e)   
...          ...          ...            ...                         ...   
2603  TUNISIENNE            3        2-LSG-5         Admis(e) - Passable   
2610  TUNISIENNE            5     2-DSI-A3-2             Admis(e) - Bien   
2611  TUNISIENNE            4     3LBC-BIS-1             Admis(e) - Bien   
2613  TUNISIENNE            5         1-BA-4         Admis(e) - Passable   
2615  TUNISIENNE            5         1-BA-5  Admis(e) - Missing credits   

      Highschool graduation grade  Final score Region  
0                             NaN         -1.0    NaN  
3                            12.0         -1.0    NaN  
10                           12.0         -1.0    NaN  
11                            NaN         -1.0    NaN  
15                           10.0         -1.0    NaN  
...                           ...          ...    ...  
2603                         10.0         -1.0    NaN  
2610              

In [20]:
df.isna().sum()

ID                                0
Birth date                        0
Birth place                       2
Highschool graduation date      362
Highschool                      164
Studies level                   362
Degree                            0
ESB entry date                 1233
Bac type                         58
Current class                     0
Semester 1 class                312
Degree level                      0
Gender                            0
Nationality                       2
Class level                       0
Previous class                  908
Previous year result            936
Highschool graduation grade      99
Final score                       0
Region                          446
dtype: int64

In [21]:
df['Birth place'].mode()

0    Tunis
Name: Birth place, dtype: object

In [22]:
# fill missing burth places with Tunis, the most recurrent place across students.
import statistics
df['Birth place'] = df['Birth place'].fillna(statistics.mode(df['Birth place']))

In [23]:
df.isna().sum()

ID                                0
Birth date                        0
Birth place                       0
Highschool graduation date      362
Highschool                      164
Studies level                   362
Degree                            0
ESB entry date                 1233
Bac type                         58
Current class                     0
Semester 1 class                312
Degree level                      0
Gender                            0
Nationality                       2
Class level                       0
Previous class                  908
Previous year result            936
Highschool graduation grade      99
Final score                       0
Region                          446
dtype: int64

In [24]:
#fill highschool graduation date based on the assumption that the student never failed any class since graduation. Example: If current class is M1, meaning current level is 4, we assum that he graduated from highschool 4 years prior to when current level school year ends: 2025 - 4
df['Highschool graduation date'] = df['Highschool graduation date'].fillna(2025 - df['Class level'])


In [25]:
df[(df['Region'].isna())& (~df['Birth place'].isin(df['Region'].unique())) & (df['Nationality'] != 'TUNISIENNE')]

ID  Birth date                   Birth place  \
23      ID24        1999                        Bongor   
62      ID63        2003                        Zinder   
461    ID462        2003                    Mauritania   
651    ID652        2001                       Yaounde   
652    ID653        2003                      Baleveng   
780    ID781        2004             Salon De Provence   
973    ID974        2001                        Bamako   
974    ID975        2007                        Bamako   
980    ID981        2003                         Diffa   
1010  ID1011        2003             Brazzaville/Congo   
1017  ID1018        2001                         R'Kiz   
1059  ID1060        1975                        Beirut   
1165  ID1166        2002                        Arafat   
1220  ID1221        2001                   Ouagadougou   
1351  ID1352        1998                  Maradi-Zaria   
1537  ID1538        2000                   Kissidougou   
1671  ID1672        2002                       Yaounde   
1680  ID1681        1999                     Bafoussam   
1771  ID1772        1999                        Douala   
1891  ID1892        2001             Douala - Cameroun   
1957  ID1958        2006                       Yaounde   
2008  ID2009        2001                 Tevragh Zeina   
2010  ID2011        1992                     Bafoussam   
2028  ID2029        1999                     N’Djamena   
2091  ID2092        1999                      Mbalmayo   
2355  ID2356        2002                        Abalak   
2357  ID2358        2001       Djeddah Arabie Saoudite   
2427  ID2428        1996                          Lomé   
2472  ID2473        2005  Adjame Abidjan Cote D Ivoire   
2543  ID2544        2003     Jacqueville Cote D'Ivoire   

      Highschool graduation date  \
23                          2021   
62                          2024   
461                         2022   
651                         2021   
652                         2024   
780                         2022   
973                         2020   
974                         2024   
980                         2022   
1010                        2022   
1017                        2020   
1059                        1999   
1165                        2021   
1220                        2022   
1351                        2020   
1537                        2018   
1671                        2022   
1680                        2017   
1771                        2016   
1891                        2020   
1957                        2024   
2008                        2019   
2010                        2016   
2028                        2018   
2091                        2020   
2355                        2021   
2357                        2019   
2427                        2013   
2472                        2022   
2543                        2022   

                                             Highschool  Studies level Degree  \
23                                      Lycee De Bongor            NaN     BA   
62                                          Lycee Niger            0.0    LBC   
461                                    Lycee Nouakchott            1.0    LSG   
651                                     Lycee De Minkan            NaN     BA   
652                                        Lycee Douala            0.0    LSG   
780                                        Lycee France            1.0    LSG   
973                                        Lycee Bamako            3.0     BA   
974                                        Lycee Bamako            0.0    LSG   
980                                         Lycee Diffa            0.0    LBC   
1010               Lycee Fondation Guillaumebrazzaville            0.0    LSG   
1017                                        Lycee R'Kiz            3.0    CCA   
1059                                       Lycee Beirut            3.0   MDSI   
1165                                   Ly

In [26]:
# Fill region values with a series of conditions
# 1. If Birth place is in existing regions, use Birth place
# 2. If Tunisian and Birth place not in regions, use 'Tunis'
# 3. If not Tunisian, use 'Foreign'

def fill_region(row):
    if pd.isna(row['Region']):
        if row['Birth place'] in df['Region'].unique():
            return row['Birth place']
        elif row['Nationality'] == 'TUNISIENNE':
            return 'Tunis'
        else:
            return 'Foreign'
    return row['Region']

df['Region'] = df.apply(fill_region, axis=1)

In [27]:
# We assumer that students with missing highschool values went to highschools in their regions:

df['Highschool'] = df['Highschool'].fillna('Lycee ' + df['Region'])


In [28]:
df.head(10)

ID  Birth date Birth place  Highschool graduation date  \
0   ID1        1996       Tunis                        2020   
1   ID2        2002       Tunis                        2021   
2   ID3        2003       Tunis                        2024   
3   ID4        1999       Tunis                        2021   
4   ID5        1999     Bizerte                        2020   
5   ID6        2004       Tunis                        2024   
6   ID7        2004       Tunis                        2024   
7   ID8        2002       Gafsa                        2022   
8   ID9        2003  Benguerden                        2022   
9  ID10        2004        Sfax                        2023   

                     Highschool  Studies level Degree  ESB entry date  \
0                   Lycee Tunis            NaN     BA            <NA>   
1                 Lycee Ezzahra            3.0    MKD            2024   
2          Lycee Asdrubal Bardo            0.0    LSG            2024   
3        Lycee Les Pères Blancs            NaN   MDSI            <NA>   
4           Lycee Bhira Bizerte            0.0    LSG            2023   
5              Lycee El Ghazela            0.0    LSG            2024   
6  Lycee El Ahdeth El Mourouj 6            0.0    LBC            2024   
7                  Lycee Nabeul            0.0    LBC            <NA>   
8               Lycee Bengarden            0.0    LBC            <NA>   
9             Lycee Borj Cedria            0.0    LBC            2024   

           Bac type Current class Semester 1 class  Degree level Gender  \
0               NaN        2-BA-2           2-BA-2             2      M   
1  Bac Informatique       1-MKD-1              NaN             1      M   
2      Bac Economie       1-LSG-1          1-LSG-1             1      M   
3      Bac Economie      1-MDSI-1         1-MDSI-1             1      F   
4      Bac Economie       2-LSG-3          2-LSG-3             2      M   
5      Bac Economie       1-LSG-5          1-LSG-5             1      M   
6      Bac Economie       1-LBC-8          1-LBC-8             1      M   
7      Bac Economie    3LBC-BIS-1       3LBC-BIS-1             3      M   
8      Bac Economie    3LBC-BIS-5       3LBC-BIS-5             3      F   
9  Bac Informatique       1-LBC-1          1-LBC-1             1      F   

  Nationality  Class level Previous class Previous year result  \
0  TUNISIENNE            5         1-BA-2             Admis(e)   
1  TUNISIENNE            4            NaN                  NaN   
2  TUNISIENNE            1            NaN                  NaN   
3  TUNISIENNE            4        3-LSG-4      Admis(e) - Bien   
4  TUNISIENNE            2        1-LSG-8  Admis(e) - Passable   
5  TUNISIENNE            1            NaN                  NaN   
6  TUNISIENNE            1            NaN                  NaN   
7  TUNISIENNE            3     2-LBC-BIS6      Admis(e) - Bien   
8  TUNISIENNE            3     2-LBC-BIS5  Admis(e) - Passable   
9  TUNISIENNE            1            NaN                  NaN   

   Highschool graduation grade  Final score     Region  
0                          NaN         -1.0      Tunis  
1                        12.91         -1.0  Ben Arous  
2                         9.81         -1.0     Ariana  
3                        12.00         -1.0      Tunis  
4                        10.48         -1.0    Bizerte  
5                        10.25         -1.0     Ariana  
6                        10.07         -1.0  Ben Arous  
7                        11.90         -1.0     Nabeul  
8                        11.34         -1.0     Ariana  
9                        10.10         -1.0  Ben Arous

In [29]:
df.groupby('Degree')['Studies level'].value_counts()

Degree  Studies level
BA      3.0               241
        0.0                 8
        1.0                 3
CCA     3.0                 9
        0.0                 1
FIND    3.0                37
        0.0                 1
LBC     0.0              1069
        1.0                37
        2.0                 2
        3.0                 1
LMAD    0.0                 6
        1.0                 1
LSG     0.0               590
        1.0                36
        3.0                 3
MDSI    3.0               137
        0.0                 8
        1.0                 4
MKD     3.0                64
        0.0                 1
Name: count, dtype: int64

In [30]:
# missing values in the Studies level column are filled with the most common value within each Degree group. If no common value exists, the missing value is nan.
df['Studies level'] = df.groupby('Degree')['Studies level'].transform(lambda x: x.fillna(x.mode()[0] if not x.mode().empty else np.nan))

In [31]:
df['Studies level'].isna().sum()

0

##### "ESB entry date" column, the number of null values is the most significant (1233). We will fill missing values based on the logic that a bachelor student enters ESB the same year they graduated highschool. Same logic is applied for first year master students, with the condition that 'Previous year result' column's value includes 'Admis', therefore minimizing inaccuracy. For second year master students, we will consider they entered ESB at their first master year.

In [32]:
# Fill missing entry date values for bachelor students
mask_licence = (
    (df['ESB entry date'].isna()) & (df['Class level'].isin([1, 2, 3])) & (~df['Previous year result'].str.contains('Redouble', na=False)) )

df.loc[mask_licence, 'ESB entry date'] = df['Highschool graduation date']

# Fill missing entry date values for master students
mask_master = (
    (df['ESB entry date'].isna()) & df['Class level'].isin([4, 5]) & (~df['Previous year result'].str.contains('Redouble', na=False)) )

df.loc[mask_master, 'ESB entry date'] = df.loc[mask_master, 'Class level'].map({4: 2021, 5: 2023})


# For remaining missing values, use the median entry date for the same degree
df['ESB entry date'] = df.groupby('Degree')['ESB entry date'].transform(lambda x: x.fillna(x.median()))

In [33]:
df['ESB entry date'].isna().sum()

0

In [34]:
df.isna().sum()

ID                               0
Birth date                       0
Birth place                      0
Highschool graduation date       0
Highschool                       0
Studies level                    0
Degree                           0
ESB entry date                   0
Bac type                        58
Current class                    0
Semester 1 class               312
Degree level                     0
Gender                           0
Nationality                      2
Class level                      0
Previous class                 908
Previous year result           936
Highschool graduation grade     99
Final score                      0
Region                           0
dtype: int64

In [35]:
#We fill missing bac type values with the most recurrent bac type in the degree program the student is enrolled in
df['Bac type'] = df['Bac type'].fillna(df.groupby('Degree')['Bac type'].transform(lambda x: x.mode().iloc[0]))


In [36]:
df.groupby('Bac type')['Highschool graduation grade'].median()

Bac type
Autre Bac           10.08
Bac Economie        10.61
Bac Français        11.87
Bac Informatique    10.61
Bac Lettres         10.87
Bac Math            11.06
Bac Sciences        11.00
Bac Technique       10.70
Name: Highschool graduation grade, dtype: float64

In [37]:
# filling missing values in the 'Highschool graduation grade' column using the median grades of students that correspond with the same Bac type. Example: A Bac Math student will have his missing highschool graduation grade filled by the median of grades of all students holding a bac math.
df['Highschool graduation grade'] = df['Highschool graduation grade'].fillna(df.groupby('Bac type')['Highschool graduation grade'].transform('median'))

In [38]:
df[(df['Current class'] != df['Semester 1 class']) & (df['Semester 1 class'].notna())]

ID  Birth date   Birth place  Highschool graduation date  \
164  ID165        2000      La Marsa                        2019   
312  ID313        2000  Menzel Temim                        2019   

                        Highschool  Studies level Degree  ESB entry date  \
164  Lycee Carthage Hannibal Tunis            3.0     BA            2024   
312                    Lycee Smida            3.0     BA            2024   

             Bac type Current class Semester 1 class  Degree level Gender  \
164  Bac Informatique        1-BA-1           1-BA-7             1      M   
312  Bac Informatique        1-BA-4           1-BA-6             1      F   

    Nationality  Class level Previous class Previous year result  \
164  TUNISIENNE            4            NaN                  NaN   
312  TUNISIENNE            4            NaN                  NaN   

     Highschool graduation grade  Final score Region  
164                        12.93         -1.0  Tunis  
312                        10.36         -1.0  Tunis

In [39]:
# we are in semester 1. Assumption that Semester 1 class = Current class

df['Semester 1 class'] = df['Semester 1 class'].combine_first(df['Current class'])

In [40]:
df.isna().sum()

ID                               0
Birth date                       0
Birth place                      0
Highschool graduation date       0
Highschool                       0
Studies level                    0
Degree                           0
ESB entry date                   0
Bac type                         0
Current class                    0
Semester 1 class                 0
Degree level                     0
Gender                           0
Nationality                      2
Class level                      0
Previous class                 908
Previous year result           936
Highschool graduation grade      0
Final score                      0
Region                           0
dtype: int64

In [41]:
df[df['Nationality'].isna()]

ID  Birth date Birth place  Highschool graduation date  \
23    ID24        1999      Bongor                        2021   
651  ID652        2001     Yaounde                        2021   

          Highschool  Studies level Degree  ESB entry date      Bac type  \
23   Lycee De Bongor            3.0     BA            2021  Bac Sciences   
651  Lycee De Minkan            3.0     BA            2021  Bac Sciences   

    Current class Semester 1 class  Degree level Gender Nationality  \
23         1-BA-3           1-BA-3             1      M         NaN   
651        1-BA-7           1-BA-7             1      F         NaN   

     Class level Previous class Previous year result  \
23             4        3-LSG-1  Admis(e) - Passable   
651            4        3-LSG-1      Admis(e) - Bien   

     Highschool graduation grade  Final score   Region  
23                          11.0         -1.0  Foreign  
651                         10.0         -1.0  Foreign

In [42]:
df.loc[df['Nationality'].isna(), 'Nationality'] = ['TCHADIENNE', 'CAMEROUNAISE']

In [43]:
# Wwe fill previous class empty fields by substracting 1 from the current class digit indicating the level. Example : Current class 2-BA-8. Assuming that student remains in the same group, we substract 1 from the first digit (2). PRevious class becomes therefroe : 1-BA-8
def fill_previous_class(row):
    if pd.isna(row['Previous class']):
        current_class = row['Current class']
        
        # Ensure the first character is numerical and properly subtract 1
        if current_class and current_class[0].isdigit():
            if int(current_class[0]) > 1:
                # Subtract 1 only from the first digit of the current class and attach rest of value
                previous_class = str(int(current_class[0]) - 1) + current_class[1:]
            else:
                # If the first digit is 1, return 'Unknown'
                previous_class = 'External'
        else:
            # Handle unexpected values safely
            previous_class = 'External'
        
        return previous_class
    # If the 'Previous class' already has a value, keep as is
    return row['Previous class']

df['Previous class'] = df.apply(fill_previous_class, axis=1)


In [44]:
df.isna().sum()

ID                               0
Birth date                       0
Birth place                      0
Highschool graduation date       0
Highschool                       0
Studies level                    0
Degree                           0
ESB entry date                   0
Bac type                         0
Current class                    0
Semester 1 class                 0
Degree level                     0
Gender                           0
Nationality                      0
Class level                      0
Previous class                   0
Previous year result           936
Highschool graduation grade      0
Final score                      0
Region                           0
dtype: int64

In [45]:
df['Previous year result'] = df.apply(
    lambda row: 'Redouble' if pd.isna(row['Previous year result']) and row['Current class'] == row['Previous class'] else ('Admis(e)' if pd.isna(row['Previous year result']) else row['Previous year result']), axis=1
)

In [46]:
df['Previous year result'].isna().sum()

0

In [47]:
df.isna().sum()

ID                             0
Birth date                     0
Birth place                    0
Highschool graduation date     0
Highschool                     0
Studies level                  0
Degree                         0
ESB entry date                 0
Bac type                       0
Current class                  0
Semester 1 class               0
Degree level                   0
Gender                         0
Nationality                    0
Class level                    0
Previous class                 0
Previous year result           0
Highschool graduation grade    0
Final score                    0
Region                         0
dtype: int64

In [48]:
df['Previous class'].value_counts()

Previous class
External      887
1-LBC-1        34
2-LBC-BI-2     33
1-LBC-3        31
2-LSG-4        31
             ... 
3LBC-BIS-1      5
3-LMAD-1        4
1-GAMMA-1       3
2-MKD-1         1
1-LSC-2         1
Name: count, Length: 86, dtype: int64

In [49]:
df.isna().sum()

ID                             0
Birth date                     0
Birth place                    0
Highschool graduation date     0
Highschool                     0
Studies level                  0
Degree                         0
ESB entry date                 0
Bac type                       0
Current class                  0
Semester 1 class               0
Degree level                   0
Gender                         0
Nationality                    0
Class level                    0
Previous class                 0
Previous year result           0
Highschool graduation grade    0
Final score                    0
Region                         0
dtype: int64

In [50]:
df['Previous year result'].value_counts()


Previous year result
Admis(e)                      1114
Admis(e) - Passable            668
Admis(e) - Bien                575
Admis(e) - Missing credits     160
Redouble                        92
Exclu(e)                        12
Name: count, dtype: int64

In [51]:
df.to_csv('MLDATA.csv', index=False)  
